### PR TITLE
[Play] - News Feed Application 

### PR DESCRIPTION
**Pull request by NagarjunShroff**
> First thing, PLEASE READ THIS: [ReactPlay Code Review Checklist](https://github.com/reactplay/react-play/wiki/ReactPlay-Code-Review-Checklist)

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)
#670 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. 

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by murtuzaalisurti** [comment]
@atapas @koustov  here email is shown instead of name, is it because of github or user entry?

chrome_AXkCHDBcEY

**Review by murtuzaalisurti — CHANGES_REQUESTED** [review]
@NagarjunShroff I am getting a 426 error from the API.

chrome_wU49i7MYru

**Comment by NagarjunShroff** [comment]
> @NagarjunShroff I am getting a 426 error from the API.
> 
> chrome_wU49i7MYru

@murtuzaalisurti Checking...

**Comment by atapas** [comment]
> @atapas @koustov here email is shown instead of name, is it because of github or user entry?
> 
> chrome_AXkCHDBcEY

@NagarjunShroff Did you provide your github id while creating the play?

File: src/plays/news-feed-application/NewsFeedApplication.jsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import "./styles.css";
+
+function NewsFeedApplication(props) {
+  const [newsData, setNewsData] = useState();
+  const [selectedCountry, updateSelectedCountry] = useState("IN");
+  const [selectedCategory, updateSelectedCategory] = useState("general");
+
+  useEffect(() => {
+    async function fetchData() {
+      const res = await fetch(
```
Comment (harshsinghatz) [inline src/plays/news-feed-application/NewsFeedApplication.jsx:20]: you can directly destructure the data here

`
const { articles } 
`

File: src/plays/news-feed-application/component/Card.jsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import Card from "@mui/material/Card";
+import CardContent from "@mui/material/CardContent";
+import CardMedia from "@mui/material/CardMedia";
+
+const NewsCard = (props) => {
```
Comment (harshsinghatz) [inline src/plays/news-feed-application/component/Card.jsx:5]: Destructure the props

File: src/plays/news-feed-application/styles.css [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    grid-gap: 24px;
+    padding: 84px 16px;
+    justify-content: space-evenly;
+    width: 100%;
+    height: 100%;
+}
+
+.card {
+    cursor: pointer;
+    background-color: #fff !important;
```
Comment (harshsinghatz) [inline src/plays/news-feed-application/styles.css:13]: is !important necessary here?

File: src/plays/news-feed-application/countriesAndCategories.js [hunk 4]
```diff
@@ -0,0 +1,330 @@
+export const categories = ["business", "health", "entertainment", "science", "sports", "technology", "general"];
+
+export const countries = [
+    {
+      code: "AR",
```
Comment (harshsinghatz) [inline src/plays/news-feed-application/countriesAndCategories.js:5]: is code3 and number property getting used anywhere if not please remove extra data

File: src/plays/news-feed-application/styles.css (hunk 3 above)
Comment (NagarjunShroff) [reply src/plays/news-feed-application/styles.css:13]: yes to override the styles mui card.

**Comment by NagarjunShroff** [comment]
> > @atapas @koustov here email is shown instead of name, is it because of github or user entry?
> > chrome_AXkCHDBcEY
> 
> @NagarjunShroff Did you provide your github id while creating the play?

@atapas, yes i did

File: src/plays/news-feed-application/NewsFeedApplication.jsx [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                <div>
+                  <FormControl size="small" className="menu-form-control">
+                    <InputLabel id="country-select-label">Country</InputLabel>
+                    <Select
+                      labelId="country-select-label"
+                      id="country-select"
+                      defaultValue={"IN"}
+                      value={selectedCountry}
+                      label="Country"
+                      onChange={(e) => {
```
Comment (harshsinghatz) [inline src/plays/news-feed-application/NewsFeedApplication.jsx:54]: You can directly write this as no need for create a new func
`
onChange={handleCountryChange}
`

File: src/plays/news-feed-application/NewsFeedApplication.jsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                        handleCountryChange(e);
+                      }}
+                    >
+                      {countries.map((country) => (
```
Comment (harshsinghatz) [inline src/plays/news-feed-application/NewsFeedApplication.jsx:58]: Try destructuring these as {name,code}

File: src/plays/news-feed-application/component/Card.jsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import Card from "@mui/material/Card";
+import CardContent from "@mui/material/CardContent";
+import CardMedia from "@mui/material/CardMedia";
+
+const NewsCard = ({ news }) => {
```
Comment (harshsinghatz) [inline src/plays/news-feed-application/component/Card.jsx:5]: Try to destructure the props

File: src/plays/news-feed-application/component/Card.jsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    >
+      <CardMedia component="img" height="140" image={news.urlToImage} />
+      <CardContent>
+        <div style={{ fontSize: "16px", fontWeight: 600, lineHeight: 1.5 }}>
+          {news.title}
+        </div>
+      </CardContent>
+      <div className="footer">
+        <div className="author">
+          {news.author !== null &&
```
Comment (harshsinghatz) [inline src/plays/news-feed-application/component/Card.jsx:20]: Simply write 
`
news.author && ...
`

Comment (NagarjunShroff) [reply src/plays/news-feed-application/component/Card.jsx:20]: Api gives null for that field if is not there.

**Comment by atapas** [comment]
Let's wrap up the review and fixes for merge.

**Comment by NagarjunShroff** [comment]
> @NagarjunShroff Have we added the API key in vercel?

No we haven't added it - Could you please add it?

**Review by murtuzaalisurti — CHANGES_REQUESTED** [review]
error on load, probably because of api key

chrome_OHapEfhFmR

**Comment by atapas** [comment]
> Visit Preview

I remember adding it. I shall check again. Irrespective of it, we need to handle the error messages correctly!

Due to some reasons, the `newsData` is undefined and breaks! The reason can be anything, including there is the API keys. Let's handle the error message correctly @NagarjunShroff 


```
 <div className="card-container">
              {newsData.length > 0 ? (
                newsData.map((news, i) => <NewsCard news={news} key={i} />)
              ) : (
                <CircularProgress />
              )}
 </div>
```

**Comment by koustov** [comment]
```json
{"status":"error","code":"corsNotAllowed","message":"Requests from the browser are not allowed on the Developer plan, except from localhost."}
```
The free key cannot be used on a production server. So @NagarjunShroff what needs to be done for this play?

**Comment by NagarjunShroff** [comment]
> ```json
> {"status":"error","code":"corsNotAllowed","message":"Requests from the browser are not allowed on the Developer plan, except from localhost."}
> ```
> 
> The free key cannot be used on a production server. So @NagarjunShroff what needs to be done for this play?

@koustov, if APIs don't work- We can ignore it play.

**Comment by atapas** [comment]
@NagarjunShroff 

Why shelve the hard work?

Can you repurpose it using the free API?

Here is an example: https://saurav.tech/NewsAPI/

**Comment by NagarjunShroff** [comment]
> @NagarjunShroff
> 
> Why shelve the hard work?
> 
> Can you repurpose it using the free API?
> 
> Here is an example: https://saurav.tech/NewsAPI/

@atapas, Sure will do it.

**Comment by NagarjunShroff** [comment]
Ready for review.

**Comment by NagarjunShroff** [comment]
@atapas Please review.

**Review by atapas — APPROVED** [review]
Looks good

**Review by koustov — APPROVED** [review]
lgtm

**Pull request merged**